controller/user_blackbox_test.go

```diff
@@ -0,0 +1,138 @@
+	"github.com/goadesign/goa"
+	"github.com/goadesign/goa/middleware/security/jwt"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+)
+
+type UserTestSuite struct {
+	gormtestsupport.DBTestSuite
+}
+
+func (s *UserTestSuite) SetupTest() {
+	s.DBTestSuite.SetupTest()
+}
+
+func TestRunUserREST(t *testing.T) {
+	suite.Run(t, &UserTestSuite{DBTestSuite: gormtestsupport.NewDBTestSuite()})
+}
+
+func (s *UserTestSuite) SecuredController(user *auth.Identity) (*goa.Service, *UserController) {
+	svc, err := auth.ServiceAsUser("User-Service", user)
+	require.NoError(s.T(), err)
+	return svc, NewUserController(svc, s.Application)
+}
+func (s *UserTestSuite) UnsecuredController() (*goa.Service, *UserController) {
+	svc := goa.New("User-Service")
+	controller := NewUserController(svc, s.Application)
+	return svc, controller
+}
+
+func (s *UserTestSuite) TestShowClusterAvailableToUser() {
+
+	s.T().Run("ok", func(t *testing.T) {
+
+		t.Run("empty", func(t *testing.T) {
+			// given
+			identity := auth.NewIdentity()
+
+			// when
+			svc, userCtrl := s.SecuredController(identity)
+			_, clusters := test.ClustersUserOK(t, svc.Context, svc, userCtrl)
+
+			// then
+			assert.NotNil(t, clusters)
+			assert.Empty(t, clusters.Data)
+
+		})
+
+		t.Run("list single cluster", func(t *testing.T) {
+			// given
+			identity := auth.NewIdentity()
+
+			cl := testsupport.CreateCluster(t, s.DB)
+			identityCluster := testsupport.CreateIdentityCluster(t, s.DB, cl, &identity.ID)
+			require.NotNil(t, identityCluster)
+
+			// when
+			svc, userCtrl := s.SecuredController(identity)
+			_, clusters := test.ClustersUserOK(t, svc.Context, svc, userCtrl)
+
+			// then
+			require.NotNil(t, clusters)
+			require.NotNil(t, clusters.Data)
+
+			testsupport.AssertEqualClusterData(t, []repository.Cluster{*cl}, clusters.Data)
+		})
+
+		t.Run("list multiple cluster", func(t *testing.T) {
+			// given
+			// create user identity
+			identity := auth.NewIdentity()
+			identityID := identity.ID
+
+			expectedClusters := make([]repository.Cluster, 3)
+			// create random cluster and cluster identity for user
+			for i := range expectedClusters {
+				c := testsupport.CreateCluster(t, s.DB)
+				identityCluster := testsupport.CreateIdentityCluster(t, s.DB, c, &identityID)
+				require.NotNil(t, identityCluster)
+
+				expectedClusters[i] = *c
+			}
+
+			// when
+			svc, userCtrl := s.SecuredController(identity)
+			_, clusters := test.ClustersUserOK(t, svc.Context, svc, userCtrl)
+
+			// then
+			require.NotNil(t, clusters)
+			require.NotNil(t, clusters.Data)
+
+			testsupport.AssertEqualClusterData(t, expectedClusters, clusters.Data)
+		})
+	})
+
+	s.T().Run("internal error", func(t *testing.T) {
+
+		t.Run("missing token", func(t *testing.T) {
+			// given
+			svc, userCtrl := s.UnsecuredController()
+			// when/then
+			test.ClustersUserInternalServerError(t, svc.Context, svc, userCtrl)
```

Comment:
why do you get an `InternalServerError` if the token is missing? Shouldn't it be an `Unauthorized` response instead?

Reply:
See https://github.com/fabric8-services/fabric8-cluster/pull/33#discussion_r
In runtime we should get 401 but it will come from the JWT middleware. The problem is that in our tests we go directly to the controller. Our middlewares are not used in the tests.

Reply:
yes, and this has always been super-annoying that we can't test the controllers with the middlewares

Reply:
Yeah.. goa generated test could be improved :)